doc/website/advanced/how-optional-and-error-values-are-returned-in-iceoryx.md

```diff
@@ -0,0 +1,122 @@
+# How optional and error values are returned in iceoryx
+
+Many parts of the iceoryx C++ API follow a functional programming approach and allow the user to specify functions 
+which handle the possible cases, e.g. what should happen when data is received.
+
+This is very flexible but requires using the monadic types ``cxx::expected`` and ``cxx::optional``, which we
+introduce in the following sections.
+
+### Optional
+
+The type ``iox::cxx::optional<T>`` is used to indicate that there may or may not be a value of a specific type ``T``
+available. This is essentially the 'maybe [monad](https://en.wikipedia.org/wiki/Monad_(functional_programming))' in 
+functional programming. Assuming we have some optional (usually the result of some computation)
+
+```cpp
+optional<int> result = someComputation();
+```
+
+we can check for its value using
+```cpp
+if(result.has_value())
+{
+    auto value = result.value();
+    // do something with the value
+}
+else
+{
+    // handle the case that there is no value
+}
+```
+A shorthand to get the value is
+```cpp
+auto value = *result;
+```
+
+!!! attention
+    Accessing the value if there is no value terminates the application, so it must be checked beforehand.
+
+We can achieve the same with the functional approach by providing a function for both cases.
+
+```cpp
+result.and_then([](int& value) { /*do something with the value*/ })
+    .or_else([]() { /*handle the case that there is no value*/ });
+```
+Notice that we get the value by reference, so if a copy is desired it has to be created explicitly in the 
+[lambda](https://en.wikipedia.org/wiki/Anonymous_function#C++_(since_C++11)) or function we pass.
+
+The optional can be initialized from a value directly
+```cpp
+optional<int> result = 73;
+result = 37;
+```
+If the optional is default initialized, it is automatically set to its null value of type ``iox::cxx::nullopt_t``. 
+This can be also done directly by using the constant ``iox::cxx::nullopt``
+
+```cpp
+result = iox::cxx::nullopt;
+```
+
+For a complete list of available functions see 
+[``optional.hpp``](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_utils/include/iceoryx_utils/cxx/optional.hpp).
+The ``iox::cxx::optional`` behaves like the [``std::optional``](https://en.cppreference.com/w/cpp/utility/optional)
+except that it does not throw exceptions and has no undefined behavior.
```

Comment:
It has undefined behavior, see above. At least in the same sense as undefined behavior of dereferencing a nullptr goes (which will crash on all platforms known to me but in theory does not have to ... hence undefined). This is no different here. We do not throw exceptions though.

Reply:
When accessing a nullopt, we terminate. There is no undefined behavior. But I'm not sure if I got your point. Did I misunderstand something?